### PR TITLE
refactor(lint/noAriaUnsupportedElements): improve diagnostic 

### PR DESCRIPTION
**Pull request by vasucp1207**
## Summary
Closes https://github.com/biomejs/biome/issues/360

### CODE REVIEW

File: crates/biome_js_analyze/src/aria_analyzers/a11y/no_aria_unsupported_elements.rs [hunk 1]
```diff
@@ -120,4 +122,35 @@ impl Rule for NoAriaUnsupportedElements {
             }),
         )
     }
+
+    fn action(_ctx: &RuleContext<Self>, _state: &Self::State) -> Option<JsRuleAction> {
+        let element = _ctx.query();
+        let mut mutation = _ctx.root().begin();
+        let mut removed_attribute = "role".to_string();
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/aria_analyzers/a11y/no_aria_unsupported_elements.rs:129]: Do we really need to allocate the string?
```suggestion
        let mut removed_attribute = "role";
```

**Review by Conaclos — COMMENTED** [review]
Looks good! I left some comments :)

File: crates/biome_js_analyze/src/aria_analyzers/a11y/no_aria_unsupported_elements.rs [hunk 2]
```diff
@@ -120,4 +122,35 @@ impl Rule for NoAriaUnsupportedElements {
...
+                .name()
+                .ok()
+                .unwrap()
+                .as_jsx_name()
+                .unwrap()
+                .value_token()
+                .ok()
+                .unwrap();
+            let attribute_name = attribute_name.to_string();
+            removed_attribute = attribute_name;
```
Comment (vasucp1207) [inline crates/biome_js_analyze/src/aria_analyzers/a11y/no_aria_unsupported_elements.rs:143]: converted `remove_attribute` to `String` due to the assigning.

Comment (Conaclos) [reply crates/biome_js_analyze/src/aria_analyzers/a11y/no_aria_unsupported_elements.rs:135]: Just to be sure: If you remove all `to_string()` calls (line 129 and line 134), do the compiler complain?

**Review by Conaclos — APPROVED** [review]

File: crates/biome_js_analyze/src/aria_analyzers/a11y/no_aria_unsupported_elements.rs (hunk 2 above)
Comment (vasucp1207) [reply crates/biome_js_analyze/src/aria_analyzers/a11y/no_aria_unsupported_elements.rs:135]: Yeah it complains, there is a type difference between them.

File: crates/biome_js_analyze/src/aria_analyzers/a11y/no_aria_unsupported_elements.rs [hunk 3]
```diff
@@ -120,4 +122,27 @@ impl Rule for NoAriaUnsupportedElements {
...
+        let mut mutation = _ctx.root().begin();
+        let mut removed_attribute = "role".to_string();
+
+        for attribute in element.attributes() {
+            let attribute = attribute.as_jsx_attribute()?;
+            let attribute_name = attribute.name().ok()?.as_jsx_name()?.value_token().ok()?;
+            let attribute_name = attribute_name.to_string();
+            removed_attribute = attribute_name;
+            mutation.remove_node(attribute.clone());
+        }
```
Comment (Conaclos) [inline crates/biome_js_analyze/src/aria_analyzers/a11y/no_aria_unsupported_elements.rs:137]: I missed the fact that you remove every attribute of the elements. Some attribute can be valid (for example in the snapshot, the `charset` attribute should not be removed`.

Instead, we could search for the attribute using `find_map`. SOmething like this:

```suggestion
        let attribute = element.attributes().into_iter().find_map(|attribute| {
            let jsx_attribute = attribute.as_jsx_attribute()?;
            let attribute_name = jsx_attribute.name().ok()?.as_jsx_name()?.value_token().ok()?;
            let attribute_name = attribute_name.text_trimmed();
            (attribute_name.starts_with("aria-") || attribute_name == "role").then_some(attribute)
        })?;
        // remove attribute and extarct its name
```

Another alternative is replacing the rule state by the attribute to remove.

Comment (vasucp1207) [reply crates/biome_js_analyze/src/aria_analyzers/a11y/no_aria_unsupported_elements.rs]: Oh, thanks I thought that it only had valid aria and role attributes in the list and did not consider others.

**Pull request merged**